src/backend/optimizer/path/allpaths.c

```diff
@@ -3003,25 +3003,89 @@ recurse_push_qual(Node *setOp, Query *topquery,
 	}
 }
 
+/*
+ * query_contains_var_walker
+ *		Checks if a query contains an equal variable as the one passed via
+ *		the context. Check is done without recursing into subqueries.
+ */
+static bool
+query_contains_var_walker(Node *node, void *context)
+{
+	if (node == NULL)
+		return false;
+
+	if (IsA(node, Var))
+	{
+		Var		   *var = (Var *) node;
+		Var		   *target_var = (Var *) context;
+
+		if (equal(var, target_var))
+			return true;
+
+		return false;
+	}
+
+	return expression_tree_walker(node, query_contains_var_walker, context);
+}
+
+/*
+ * query_contains_lateral_vars
+ *		Checks if a query contains any lateral var from outer level specified by
+ *		levelsup. Check is done without recursing into subqueries.
+ */
+static bool
+query_contains_lateral_vars(Query *parse, List *lateral_vars, int levelsup)
+{
+	bool		result = false;
+	ListCell   *lc;
+
+	foreach(lc, lateral_vars)
+	{
+		Node	   *node = (Node *) lfirst(lc);
+
+		if (NULL == node || !IsA(node, Var))
```

Comment:
It is not ORCA-style! Change to
```suggestion
		if (node == NULL || !IsA(node, Var))
```

Reply:
not relevant now